src/frontend/src/optimizer/plan_node/stream_partition_compute.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use pretty_xmlish::XmlNode;
+use risingwave_common::catalog::Field;
+use risingwave_common::types::DataType;
+use risingwave_pb::stream_plan::stream_node::PbNodeBody;
+
+use super::expr_visitable::ExprVisitable;
+use super::stream::prelude::{GenericPlanRef, PhysicalPlanRef};
+use super::stream::StreamPlanRef;
+use super::utils::{childless_record, Distill};
+use super::{ExprRewritable, PlanBase, PlanTreeNodeUnary, Stream, StreamNode};
+use crate::stream_fragmenter::BuildFragmentGraphState;
+use crate::PlanRef;
+
+pub struct PartitionComputeInfo {
+    pub partition_type: DataType,
+    // (partition_field_name, partition_field_transform)
+    pub partition_fields: Vec<(String, String)>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PartitionField {
+    pub col_index: u64,
+    pub transform: String,
+}
+
+/// [`StreamPartitionCompute`] used to compute the partition of the stream in advance.
```

Comment:
Could you add more comments here? It's not clear to people without background what does "partition" and "stream" and "in advance" mean here.

It would be better to describe how it works and when it is used here.

Reply:
I have added doc in https://github.com/risingwavelabs/risingwave/blob/0e9f74a78f365bc013585399b5369c68223beb3b/src/frontend/src/optimizer/plan_node/stream_sink.rs#L57. Please let me know if this not clear enough.

Reply:
looks good